Mathlib/Analysis/Convex/Deriv.lean

```diff
@@ -365,19 +365,213 @@ lemma ConvexOn.slope_mono (hfc : ConvexOn 𝕜 s f) (hx : x ∈ s) : MonotoneOn
+    BddBelow (slope f x '' {y ∈ s | x < y}) := by
+  obtain ⟨y, hyx, hys⟩ : ∃ y, y < x ∧ y ∈ s :=
+    Eventually.exists_lt (mem_interior_iff_mem_nhds.mp hxs)
+  refine bddBelow_iff_subset_Ici.mpr ⟨slope f x y, fun y' ⟨z, hz, hz'⟩ ↦ ?_⟩
+  simp_rw [mem_Ici, ← hz']
+  refine hfc.slope_mono (interior_subset hxs) ?_ ?_ (hyx.trans hz.2).le
+  · simp [hys, hyx.ne]
+  · simp [hz.2.ne', hz.1]
+
+lemma bddAbove_slope_gt_of_mem_interior (hfc : ConvexOn 𝕜 s f) (hxs : x ∈ interior s) :
+    BddAbove (slope f x '' {y ∈ s | y < x}) := by
+  obtain ⟨y, hyx, hys⟩ : ∃ y, x < y ∧ y ∈ s :=
+    Eventually.exists_gt (mem_interior_iff_mem_nhds.mp hxs)
+  refine bddAbove_iff_subset_Iic.mpr ⟨slope f x y, fun y' ⟨z, hz, hz'⟩ ↦ ?_⟩
+  simp_rw [mem_Iic, ← hz']
+  refine hfc.slope_mono (interior_subset hxs) ?_ ?_ (hz.2.trans hyx).le
+  · simp [hz.2.ne, hz.1]
+  · simp [hys, hyx.ne']
+
 end slope
 
 namespace ConvexOn
 
 variable {S : Set ℝ} {f : ℝ → ℝ} {x y f' : ℝ}
 
+section Interior
+
+/-!
+### Left and right derivative of a convex function in the interior of the set
+-/
+
+lemma hasDerivWithinAt_sInf_slope_of_mem_interior (hfc : ConvexOn ℝ S f) (hxs : x ∈ interior S) :
+    HasDerivWithinAt f (sInf (slope f x '' {y ∈ S | x < y})) (Ioi x) x := by
+  have hxs' := hxs
+  rw [mem_interior_iff_mem_nhds, mem_nhds_iff_exists_Ioo_subset] at hxs'
+  obtain ⟨a, b, hxab, habs⟩ := hxs'
+  simp_rw [hasDerivWithinAt_iff_tendsto_slope]
+  simp only [mem_Ioi, lt_self_iff_false, not_false_eq_true, diff_singleton_eq_self]
+  have h_bddBelow : BddBelow (slope f x '' Ioo x b) := by
+    refine (bddBelow_slope_lt_of_mem_interior hfc hxs).mono ?_
+    exact image_subset _ fun z hz ↦ ⟨habs ⟨hxab.1.trans hz.1, hz.2⟩, hz.1⟩
+  have h_Ioo : Tendsto (slope f x) (𝓝[>] x) (𝓝 (sInf (slope f x '' Ioo x b))) := by
+    refine MonotoneOn.tendsto_nhdsWithin_Ioo_right ?_ ?_ h_bddBelow
+    · simpa using hxab.2
+    · exact (monotoneOn_slope_gt hfc (interior_subset hxs)).mono
+        fun z hz ↦ ⟨habs ⟨hxab.1.trans hz.1, hz.2⟩, hz.1⟩
+  suffices sInf (slope f x '' Ioo x b) = sInf (slope f x '' {y ∈ S | x < y}) by rwa [← this]
+  apply le_antisymm
+  · refine le_csInf ?_ fun z hz ↦ ?_
+    · simp only [image_nonempty]
+      obtain ⟨z, hxz, hzb⟩ := exists_between hxab.2
+      exact ⟨z, habs ⟨hxab.1.trans hxz, hzb⟩, hxz⟩
+    · simp only [mem_image, mem_setOf_eq] at hz
+      obtain ⟨y, ⟨hys, hxy⟩, rfl⟩ := hz
+      obtain ⟨z, hxz, hzy⟩ := exists_between (lt_min hxab.2 hxy)
+      refine csInf_le_of_le (b := slope f x z) h_bddBelow ?_ ?_
+      · exact ⟨z, ⟨hxz, hzy.trans_le (min_le_left _ _)⟩, rfl⟩
+      · refine monotoneOn_slope_gt hfc (interior_subset hxs) ?_ ⟨hys, hxy⟩
+          (hzy.le.trans (min_le_right _ _))
+        exact ⟨habs ⟨hxab.1.trans hxz, hzy.trans_le (min_le_left _ _)⟩, hxz⟩
+  · refine csInf_le_csInf (bddBelow_slope_lt_of_mem_interior hfc hxs) ?_ ?_
+    · simpa using hxab.2
+    · exact image_subset _ fun z hz ↦ ⟨habs ⟨hxab.1.trans hz.1, hz.2⟩, hz.1⟩
```

Comment:
In the course of this proof, you show `Ioo x b ⊆ {y | y ∈ S ∧ x < y}` three times: once on line 434, again on line 453 and once again on line 456. It seems sensible to me to extract that proof into a have statement inside this proof.